hedera-node/hedera-app/src/main/java/com/hedera/node/app/signature/hapi/SignatureVerificationFuture.java

```diff
@@ -0,0 +1,305 @@
+                millisRemaining -= System.currentTimeMillis() - now;
+                if (millisRemaining <= 0) {
+                    // If there was no time left, then TimeoutException
+                    throw new TimeoutException("Timed out waiting for signature verification to complete");
+                } else if (txSig.getFuture() == null) {
+                    // If there was time left, but we didn't yet have a Future on the TransactionSignature, then
+                    // rather than blocking in a tight loop (like we do for "get"), we have to just wait a bit and
+                    // check for timeout and try again. That way, we can still time out.
+                    Thread.sleep(1);
+                } else {
+                    // We know the maximum number of millis we can wait, so we block for at most that long. If it takes
+                    // longer, the future we delegate to here will throw. If it takes less, we recompute how much is
+                    // remaining on the next iteration.
+                    txSig.getFuture().get(millisRemaining, TimeUnit.MILLISECONDS);
+                    futureReady = true;
+                }
+            }
+        }
+
+        return new SignatureVerificationImpl(
+                key, hollowAccount, new ArrayList<>(sigs.values()), checkIfPassed(key, sigs));
+    }
+
+    /**
+     * Determines whether the signature verification as a whole was successful. Called only <strong>AFTER</strong> we
+     * have waited for all {@link TransactionSignature} {@link Future}s, so we know we have either
+     * {@link VerificationStatus#VALID} or {@link VerificationStatus#INVALID} as a final result for each key.
+     *
+     * <p>Each {@link TransactionSignature} represents the signature check for a SINGLE concrete key type. There is
+     * no {@link TransactionSignature} created for a {@link KeyList} or {@link ThresholdKey}, or for "internal" keys
+     * like for contract IDs.
+     *
+     * <p>It is possible that in a complex multi-sig threshold-key scenario that some {@link TransactionSignature}s
+     * may be {@link VerificationStatus#INVALID}, and yet the overall result is still valid. More specifically:
+     *
+     * <ol>
+     *     <li>If the key is a {@link KeyList}, every single key within it must be valid</li>
+     *     <li>If the key is a {@link ThresholdKey}, then only {@link ThresholdKey#threshold()} keys must be valid</li>
+     *     <li>For all supported primitive keys, they must be valid</li>
+     *     <li>Any unsupported primitive keys or "internal" keys are always invalid</li>
+     * </ol>
+     *
+     * @param key The key to verify
+     * @param map A Map of {@link Key} to {@link TransactionSignature}
+     * @return {@code true} if the {@code key} has been validated, {@code false} otherwise.
+     */
+    private boolean checkIfPassed(@NonNull final Key key, @NonNull final Map<Key, TransactionSignature> map) {
+        return switch (key.key().kind()) {
```

Comment:
These three mutually recursive `check*()` variants will be convenient to extract to static methods that accept a strategy interface for testing if any of the "base case" kinds of `Keys`s have valid signatures.

Reply:
If you are OK I'd like to leave this as an exercise for a future PR. I tried to factor this out several times (similar logic is also in the SignatureVerifierImpl), but I couldn't see a nice way to do it. The logic between those two places was slightly different. I thought about using a Visitor pattern approach, but somebody on the internet wisely remarked that the visitor pattern and switch expressions do the same thing but switch expressions are way less overhead!